src/libraries/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Windows/PkcsPalWindows.Encrypt.cs

```diff
@@ -69,21 +68,37 @@ public sealed unsafe override byte[] Encrypt(CmsRecipientCollection recipients,
 
         private static void ReencodeIfUsingIndefiniteLengthEncodingOnOuterStructure(ref byte[] encodedContent)
         {
-            AsnReader reader = new AsnReader(encodedContent, AsnEncodingRules.BER);
-            reader.ReadTagAndLength(out int? contentsLength, out int _);
-
-            if (contentsLength != null)
+            try
             {
-                // definite length, do nothing
-                return;
-            }
+                Asn1Tag.Decode(encodedContent, out int tagLength);
 
-            using (AsnWriter writer = new AsnWriter(AsnEncodingRules.BER))
-            {
+                if (encodedContent.Length <= tagLength)
+                {
+                    throw new CryptographicException(SR.Cryptography_Der_Invalid_Encoding);
+                }
+
+                if (encodedContent[tagLength] != 0x80)
```

Comment:
It could be done without the optimization by calling ReadEncodedValue and observing whether `contentOffset + contentLength != bytesConsumed` (to be indefinite length).

We can always add `int? AsnDecoder.ReadLength(ReadOnlySpan<byte> source, AsnEncodingRules ruleSet)` if it feels like things other than this ends up caring (though all it really wanted was the boolean half of `int?`: am I definite, or indefinite?

Reply:
Yeah, I wouldn't hold up the merge for this, especially since there's only one consumer. It just stood out as a potential gap if we think this will be a common scenario.